File: frontend/src/objects/ui/component/list-component.ts
import { z } from "zod";
import { nullishToUndefined } from "../../system/schema-utils.js";
import { ComponentBindingSchema } from "./component-binding.js";
import { ComponentPositionSchema } from "./component-position.js";
import { ComponentStyleSchema } from "./component-style.js";

export const ListComponentSchema = z.object({
  id: z.string().min(1),
  type: z.literal("list"),
  dataPath: z.string().min(1),
  itemTemplate: z.array(z.unknown()).default([]),
  emptyStateText: nullishToUndefined(z.string().min(1)),
  position: ComponentPositionSchema,
  style: nullishToUndefined(ComponentStyleSchema),
  binding: nullishToUndefined(ComponentBindingSchema),
});
export type ListComponent = z.infer<typeof ListComponentSchema>;

File: backend/microservice/src/ui/objects/component/ListComponent.scala
package microservice.ui.objects.component

/** 列表组件：dataPath 驱动，含 itemTemplate 子组件树。 */
final case class ListComponent(
  id: String,
  dataPath: String,
  itemTemplate: List[PageComponent],
  emptyStateText: Option[String],
  position: ComponentPosition,
  style: Option[ComponentStyle],
  binding: Option[ComponentBinding]
) extends PageComponent {
  override val `type`: String = "list"
}
